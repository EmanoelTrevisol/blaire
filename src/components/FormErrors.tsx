import React from 'react';
import { StyleSheet, View, Text } from 'react-native';

export interface IFormErrorProps {
  messages: {
    required: string;
    valid: string;
  };
  error: boolean;
  errors: {
    valid: boolean;
    required: boolean;
  };
  dirty: boolean;
}

const FormErrors = (props: IFormErrorProps) => {
  const {
    error,
    dirty,
    messages,
    errors: { required, valid },
  } = props;
  if (!error || !dirty) {
    return null;
  }

  return (
    <View style={stl.errors}>
      {(required && (
        <Text style={stl.inputErrorText} testID="required-error">
          {messages.required}
        </Text>
      )) ||
        null}
      {(valid && (
        <Text style={stl.inputErrorText} testID="valid-error">
          {messages.valid}
        </Text>
      )) ||
        null}
    </View>
  );
};

const stl = StyleSheet.create({
  errors: {
    marginVertical: 10,
  },
  inputErrorText: {
    color: '#dd0000',
    fontSize: 12,
  },
});

export default FormErrors;
