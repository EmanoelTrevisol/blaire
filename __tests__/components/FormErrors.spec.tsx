import React from 'react';
import FormErrors, { IProps } from '@components/FormErrors';
import { Text } from 'react-native';

import renderer from 'react-test-renderer';
let props: IProps;

describe('FormErrors.tsx', () => {
  const render = (cpProps: IProps) => {
    const comp = renderer.create(<FormErrors {...cpProps} />);
    return {
      comp,
      root: comp.root,
    };
  };

  beforeEach(() => {
    props = {
      messages: {
        required: 'Obrigatório',
        valid: 'Inválido',
      },
      dirty: true,
      error: true,
      errors: {
        valid: true,
        required: true,
      },
    };
  });

  test('matches snaptshot', () => {
    const { comp } = render(props);

    expect(comp.toJSON()).toMatchSnapshot();
  });

  test('shows required and valid error', () => {
    const { root } = render(props);
    const errors = root.findAllByType(Text);

    expect(errors.length).toBe(2);

    expect(errors[0].props.children).toBe(props.messages.required);
    expect(errors[1].props.children).toBe(props.messages.valid);
  });

  test('hides required error when updated', () => {
    const newProps = { ...props, errors: { ...props.errors } };

    newProps.errors.required = false;

    const { root, comp } = render(props);
    const errors = root.findAllByType(Text);

    expect(errors.length).toBe(2);

    comp.update(<FormErrors {...newProps} />);

    const newErrors = root.findAllByType(Text);
    expect(newErrors.length).toBe(1);

    expect(newErrors[0].props.children).not.toBe(props.messages.required);
  });

  test('hides valid error when updated', () => {
    const newProps = { ...props, errors: { ...props.errors } };

    newProps.errors.valid = false;

    const { root, comp } = render(props);
    const errors = root.findAllByType(Text);

    expect(errors.length).toBe(2);

    comp.update(<FormErrors {...newProps} />);

    const newErrors = root.findAllByType(Text);
    expect(newErrors.length).toBe(1);

    expect(newErrors[0].props.children).not.toBe(props.messages.valid);
  });

  test('renders null when dirty is false', () => {
    props.dirty = false;
    const { comp } = render(props);

    expect(comp.toJSON()).toBe(null);
  });

  test('renders null when error is false', () => {
    props.error = false;
    const { comp } = render(props);

    expect(comp.toJSON()).toBe(null);
  });

  test('hides errors when error is false', () => {
    const newProps = {
      ...props,
      error: false,
      dirty: false,
      errors: { ...props.errors, required: false, valid: false },
    };
    const { comp } = render(props);

    comp.update(<FormErrors {...newProps} />);

    expect(comp.toJSON()).toBe(null);
  });

  test('changes required and valid errors correctly when props update', () => {
    const newProps = { ...props, errors: { ...props.errors } };

    props.errors.valid = true;
    props.errors.required = false;

    newProps.errors.valid = false;
    newProps.errors.required = true;

    const { root, comp } = render(props);
    const firstErrors = root.findAllByType(Text);

    expect(firstErrors.length).toBe(1);
    expect(firstErrors[0].props.children).toBe(props.messages.valid);

    comp.update(<FormErrors {...newProps} />);

    const updatedErrors = root.findAllByType(Text);

    expect(updatedErrors.length).toBe(1);
    expect(updatedErrors[0].props.children).not.toBe(props.messages.valid);
  });
});
