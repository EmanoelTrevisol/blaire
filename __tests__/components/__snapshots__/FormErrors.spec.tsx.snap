// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormErrors.tsx matches snaptshot 1`] = `
<View
  style={
    Object {
      "marginVertical": 10,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#dd0000",
        "fontSize": 12,
      }
    }
    testID="required-error"
  >
    Obrigatório
  </Text>
  <Text
    style={
      Object {
        "color": "#dd0000",
        "fontSize": 12,
      }
    }
    testID="valid-error"
  >
    Inválido
  </Text>
</View>
`;
